File: src/PW/PW/Client/Shared/pwApiClient.ts
module PW {
	export class Duration {
		public from: moment.Moment;
		public to: moment.Moment;
	}

	export class PagingData {
		public skip: number;
		public take: number;
	}

	export interface IPwApiClient {
		getToken(username: string, password: string): ng.IHttpPromise<TokenData>;
		signIn(dto: IAuthRequestVDTO): ng.IHttpPromise<IAuthResultVDTO>;
		signUp(dto: ISignupRequestVDTO): ng.IHttpPromise<IAuthResultVDTO>;
		getTransfers(userId: string, duration: Duration, paging: PagingData, token: string): ng.IHttpPromise<ITransferVDTO[]>;
		getTransfersByUserTo(userId: string, usernameTo: string, duration: Duration, paging: PagingData, token: string): ng.IHttpPromise<ITransferVDTO[]>;
		getTransfersCount(userId: string, duration: Duration, token: string): ng.IHttpPromise<number>;
		getTransfersCountByUserTo(userId: string, usernameTo: string, duration: Duration, token: string): ng.IHttpPromise<number>;
		getUserById(id: string): ng.IHttpPromise<IUserVDTO>;
		getUsers(): ng.IHttpPromise<IUserVDTO[]>;
		getUsersBySearchText(searchText: string): ng.IHttpPromise<IUserVDTO[]>;
	}

	export class TokenResponseVDTO {
		public data: TokenData;
	}

	export class TokenData {
		public access_token: string;
		public expires_in: number;
		public token_type: string;
		
	}



	export class PwApiClient implements IPwApiClient {

		public static get DATE_FORMAT(): string { return "YYYY-MM-DD HH:mm"; }

		constructor(private $http: ng.IHttpService, private $resource: ng.resource.IResourceService, private securityProvider: ISecurityProvider) {
		}

		public getToken(username: string, password: string): ng.IHttpPromise<TokenData> {
			this.$http.defaults.headers.post['Content-Type'] = "application/x-www-form-urlencoded";
			return this.$http({
				url: "/token",
				method: "POST",
				data: `grant_type=password&username=${username}&password=${password}`,
				file: {}
			});
			
		}

		public signIn(dto: IAuthRequestVDTO): ng.IHttpPromise<IAuthResultVDTO> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";
			return this.$http.post(`/api/auth`, dto);
		}

		public signUp(dto: ISignupRequestVDTO): ng.IHttpPromise<IAuthResultVDTO> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";
			return this.$http.post(`/api/auth/signup`, dto);
		}

		public getTransfers(userId: string, duration: Duration, paging: PagingData, token: string): ng.IHttpPromise<ITransferVDTO[]> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";

			return this.$http.get(`/api/transfer?userId=${userId}
					&from=${duration.from.format(PwApiClient.DATE_FORMAT)}&to=${duration.to.format(PwApiClient.DATE_FORMAT)}
					&skip=${paging.skip}&take=${paging.take}`,
					{headers: { "Authorization": "Bearer " + token}});
		}

		public getTransfersByUserTo(userId: string, usernameTo: string, duration: Duration, paging: PagingData, token:string): ng.IHttpPromise<ITransferVDTO[]> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";

			return this.$http.get(`/api/transfer?userId=${userId}&usernameTo=${usernameTo}
					&from=${duration.from.format(PwApiClient.DATE_FORMAT)}&to=${duration.to.format(PwApiClient.DATE_FORMAT)}
					&skip=${paging.skip}&take=${paging.take}`,
					{headers: { "Authorization": "Bearer " + token } });
		}

		public getTransfersCount(userId: string, duration: Duration, token: string): ng.IHttpPromise<number> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";


			return this.$http.get(`/api/transfer/count?userId=${userId}
					&from=${duration.from.format(PwApiClient.DATE_FORMAT)}&to=${duration.to.format(PwApiClient.DATE_FORMAT)}`,
				{headers: { "Authorization": "Bearer " + token } });
		}

		public getTransfersCountByUserTo(userId: string, usernameTo: string, duration: Duration, token: string): ng.IHttpPromise<number> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";


			return this.$http.get(`/api/transfer/count?userId=${userId}&usernameTo=${usernameTo}
					&from=${duration.from.format(PwApiClient.DATE_FORMAT)}&to=${duration.to.format(PwApiClient.DATE_FORMAT)}`,
				{headers: { "Authorization": "Bearer " + token } });
		}

		public getUserById(id: string): ng.IHttpPromise<IUserVDTO> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";

			return this.$http.get(`/api/user?id=${id}`);
		}

		public getUsers(): ng.IHttpPromise<IUserVDTO[]> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";
			return this.$http.get(`/api/user`);
		}

		public getUsersBySearchText(searchText: string): ng.IHttpPromise<IUserVDTO[]> {
			this.$http.defaults.headers.post['Content-Type'] = "application/json";
			return this.$http.get(`/api/user?searchText=${searchText}`);
		}
	}

	angular.module("PW")
		.factory("PW.PwApiClient", ["$http", "$resource", "PW.SecurityProvider",
			($http, $resource, securityProvider) => new PwApiClient($http, $resource, securityProvider)]);

}

File: src/PW.Mobile/PW.Mobile.API/PwApiClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;
using PW.Mobile.API.DTO;
using PW.Mobile.Infrastructure.Services;

namespace PW.Mobile.API
{
	public interface IPwApiClient
	{
		//Task<AuthResultVDTO> AuthorizeAsync(AuthRequestVDTO dto);
		Task<AuthResultVDTO> SignUp(SignupRequestVDTO dto);
		Task<UserVDTO[]> GetUsersAsync(int skip, int take);
		Task<UserVDTO[]> GetUsersAsync(string searchText);
		Task<TransferVDTO[]> GetTransfersAsync(Guid userId, DateTime from, DateTime to, string accessToken);
		Task<TokenResponse> GetTokenAsync(string username, string password);
	}

	public class TokenResponse
	{
		[JsonProperty("access_token")]
		public string AccessToken { get; set; }
		[JsonProperty("refresh_token")]
		public string RefreshToken { get; set; }
	}

	public class PwApiClient : IPwApiClient
	{
		private readonly string _baseAddress;
		private readonly ISecurityProvider _securityProvider;

		public PwApiClient(string baseAddress, ISecurityProvider securityProvider)
		{
			_baseAddress = baseAddress;
			_securityProvider = securityProvider;
		}

		public async Task<TokenResponse> GetTokenAsync(string username, string password)
		{
			using (var client = new HttpClient())
			{
				client.BaseAddress = new Uri(_baseAddress);
				var rawResult = await client.PostAsync("/token",
					new FormUrlEncodedContent(new[]
					{
						new KeyValuePair<string, string>(
							"grant_type",
							"password"),

						new KeyValuePair<string, string>(
							"scope",
							"openid profile offline_access"),

						new KeyValuePair<string, string>(
							"username",
							username),

						new KeyValuePair<string, string>(
							"password",
							password),
					}));

				var data = await rawResult.Content.ReadAsStringAsync();

				return JsonConvert.DeserializeObject<TokenResponse>(data);
			}
		}

		public async Task<AuthResultVDTO> SignUp(SignupRequestVDTO dto)
		{
			var json = JsonConvert.SerializeObject(dto);
			return await PostAsync<AuthResultVDTO>("/api/auth/signup", new StringContent(json, Encoding.UTF8, "application/json"));
		}

		public async Task<UserVDTO[]> GetUsersAsync(int skip, int take)
		{
			return await GetAsync<UserVDTO[]>($"/api/user?skip={skip}&take={take}");

		}

		public async Task<UserVDTO[]> GetUsersAsync(string searchText)
		{
			return await GetAsync<UserVDTO[]>($"/api/user?searchText={searchText}");
		}

		public async Task<TransferVDTO[]> GetTransfersAsync(Guid userId, DateTime from, DateTime to, string accessToken)
		{
			using (var client = new HttpClient())
			{
				client.BaseAddress = new Uri(_baseAddress);
				client.DefaultRequestHeaders.Accept.Clear();
				client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
				client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);

				var requestUri = $"/api/transfer?userId={userId}" +
				                    $"&from={@from.ToString("yyyy-MM-dd HH:mm")}" +
				                    $"&to={to.ToString("yyyy-MM-dd HH:mm")}";

				client.BaseAddress = new Uri(_baseAddress);
				var result = await client.GetAsync(requestUri);

				var rawResponse = result.Content.ReadAsByteArrayAsync().Result;
				var jsonResult = Encoding.UTF8.GetString(rawResponse, 0, rawResponse.Length);
				var authResult = JsonConvert.DeserializeObject<TransferVDTO[]>(jsonResult);

				return authResult;
			}
		}

		private async Task<T> PostAsync<T>(string requestUri, HttpContent content)
		{
			using (var client = new HttpClient())
			{
				client.BaseAddress = new Uri(_baseAddress);
				client.DefaultRequestHeaders.Accept.Clear();
				client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

				var result = await client.PostAsync(requestUri, content);

				var rawResponse = result.Content.ReadAsByteArrayAsync().Result;
				var jsonResult = Encoding.UTF8.GetString(rawResponse, 0, rawResponse.Length);
				var authResult = JsonConvert.DeserializeObject<T>(jsonResult);

				return authResult;
			}
		}

		private async Task<T> GetAsync<T>(string requestUri)
		{
			using (var client = new HttpClient())
			{
				client.BaseAddress = new Uri(_baseAddress);
				client.DefaultRequestHeaders.Accept.Clear();
				client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

				var result = await client.GetAsync(requestUri);

				var rawResponse = result.Content.ReadAsByteArrayAsync().Result;
				var jsonResult = Encoding.UTF8.GetString(rawResponse, 0, rawResponse.Length);
				var authResult = JsonConvert.DeserializeObject<T>(jsonResult);

				return authResult;
			}
		}
	}
}